### ISSUE TITLE
[DISCUSS] Compute Shader non-buffer arguments

### ISSUE THREAD

**Issue by tqchen** [gpuweb/gpuweb#762]
Hi WebGPU community:

I would like to use this issue to call for a discussion about what is the best way to compile compute shader kernels with non-buffer arguments.

Compute shader is one of the exciting features that WebGPU would enable, and this would allow running efficient machine learning on the platforms that webgpu support.

One of the common pattern in compute shaders(e.g. cuda) are the presence of non-buffer arguments. For example in the following code (written in CUDA). The argument `n` is the non-buffer argument that we use to specify the length of the array.

```c
__global__
void add(float *x, float *y, int n)
{
  for (int i = 0; i < n; i++)
      y[i] = x[i] + y[i];
}
```
Here the float pointer `x` and `y` can be perfectly supported by the storage buffer. So our remaining question is how to pass n to the GPU kernel.

While in this particular case, `n` is not necessary to be passed to the kernel, since we could deduce it from the size of the buffer. However, it is no longer the case as we start to work with higher dimensional tensors(e.g. conv2d that takes 4-dimensional inputs), these shape argument would need to be part of the kernel arguments.

I would like to use this thread to discuss what was the best way to support passing those additional arguments. In our previous approach via Metal or Vulkan, the additional arguments are passed via PushConstant (or setBytes in Metal's case), and they are pushed to the device in an efficient manner.

It would be great to see if push constant can become part of webgpu. Alternatively, it would be good to get thoughts about possible alternatives.
So far the only possible way seems to be create another buffer on the fly and push it to the gpu, or use a ring buffer. Although the ring buffer approach puts somewhat more burdens to the runtime developers, and I am not sure if small push constants like that could be further optimized by the underlying driver.

**Comment by tqchen** [comment]
This is related to https://github.com/gpuweb/gpuweb/issues/75, but more on the application side rather than the implementation details. So please feel free to suggest to close this one and move to the techinical discussion issue.

**Comment by kvark** [comment]
I think push constants could be a good feature to have. It's not yet clear how to roll it in. And we haven't got a sign of them giving an advantage on native, at least. Maybe in the near future we'll try to make a benchmark for push constants in the browser and see if it makes more of a difference.

The easiest way to get what you want today would be having a uniform buffer. You can update its contents before every submission, using `GPUQueue.writeBuffer`. There is no need to ring buffer them or create new buffers (any more ).

**Comment by tqchen** [comment]
I see, I wonder whether that would create the problem of write after write dep if the same function is invoked multiple times consecutively

**Comment by kvark** [comment]
Yes, `writeBuffer` is a queue operation, so the order is clear with regards to submissions. WebGPU guarantees that the following submits will observe the result of your writes. If you want to have multiple dispatches, you'd probably have a larger buffer being updated, that contains multiple uniform buffer instances in it.

**Comment by Twinklebear** [comment]
Push constants would be great to have. Sending just a size as a constant vs. a buffer is maybe a smaller hassle but it makes the API kind of painful for more complicated tasks.

For example, I needed a stream compaction operation to run on larger arrays (larger than a single dispatch), and so I have to chunk the calls up. I can use dynamic buffer offsets to specify the input position starting offsets, and I have an array of offsets telling me where to write the outputs. However the outputs I'm writing are the IDs of the active elements, which I don't have in another buffer (since it's just `gl_GlobalInvocationID.x`). So as the different chunked passes are run, the thread's `gl_GlobalInvocationID.x` needs an offset applied, based on which offset of values I'm processing. This would be easy to set through a push constant, but since this isn't supported I have to send them through a buffer. That itself would not be a big deal, but it seems like the dynamic offsets are required to be aligned to some value (e.g., in Dawn it's 256b), so each 4 byte offset I send needs an additional 254 bytes of padding so the offsets can be aligned. If I could just set a push constant each pass, this would be really convenient and a lot easier to implement. This is used as part of a data-parallel [marching cubes](https://github.com/Twinklebear/webgpu-experiments/blob/master/js/compute_marching_cubes.js).

For reference, the [shader](https://github.com/Twinklebear/webgpu-experiments/blob/master/shaders/stream_compact.comp) is:

```glsl
#version 450 core

layout(local_size_x = 1, local_size_y = 1, local_size_z = 1) in;

layout(set = 0, binding = 0) buffer Input {
    uint inputs[];
};

layout(set = 0, binding = 1) buffer Offsets {
    uint offsets[];
};

// The compaction execution offset chunk we're running
layout(set = 0, binding = 2) uniform CompactionOffset {
    uint compact_offset;
};

layout(set = 0, binding = 3) buffer Output {
    uint outputs[];
};

void main(void) {
    // Note: this is just for compacting down id's of "active" elements, so 0's are inherently
    // things we don't want to output.
    if (inputs[gl_GlobalInvocationID.x] != 0) {
        outputs[offsets[gl_GlobalInvocationID.x]] = gl_GlobalInvocationID.x + compact_offset;
    }
}
```

And the [compaction pass](https://github.com/Twinklebear/webgpu-experiments/blob/master/js/compute_marching_cubes.js#L286-L408) run as (omitting remainder handling):

```js
    // Fill our "push constants" offset index buffer
    var numChunks = Math.ceil(voxelsToProcess / this.maxDispatchSize);
    var [compactPassOffset, mapping] = this.device.createBufferMapped({
        size: numChunks * 1024,
        usage: GPUBufferUsage.UNIFORM
    });
    {
        var map = new Uint32Array(mapping);
        for (var i = 0; i < numChunks; ++i) {
            map[i * 64] = i * this.maxDispatchSize;
        }
        compactPassOffset.unmap();
    }

    var streamCompactBG = this.device.createBindGroup({
        layout: this.streamCompactLayout,
        entries: [
            {
                binding: 0,
                resource: {
                    buffer: this.voxelActiveBuffer,
                    size: 4 * Math.min(voxelsToProcess, this.maxDispatchSize),
                    offset: 0
                }
            },
            {
                binding: 1,
                resource: {
                    buffer: this.activeVoxelOffsets,
                    size: 4 * Math.min(voxelsToProcess, this.maxDispatchSize),
                    offset: 0
                }
            },
            {
                binding: 2,
                resource: {
                    buffer: compactPassOffset,
                    size: 4,
                    offset: 0
                }
            },
            {
                binding: 3,
                resource: {
                    buffer: activeVoxelIds
                }
            }
        ]
    });

    var commandEncoder = this.device.createCommandEncoder();
    var pass = commandEncoder.beginComputePass();
    pass.setPipeline(this.streamCompactPipeline);
    for (var i = 0; i < numChunks; ++i) {
        var numWorkGroups = Math.min(voxelsToProcess - i * this.maxDispatchSize, this.maxDispatchSize);
        var offset = i * this.maxDispatchSize * 4;
        pass.setBindGroup(0, streamCompactBG, [offset, offset, i * 256]);
        pass.dispatch(numWorkGroups, 1, 1);
    }
    pass.endPass();
    this.device.defaultQueue.submit([commandEncoder.finish()]);

**Comment by dneto0** [comment]
Our OpenCL C to Vulkan compute shader (https://github.com/google/clspv) compiler supports 3 ways for passing plain-old-data (POD) arguments:
- buffer per argument
- collect them into a single UBO
- collect them into a push-constant variable

We're still measuring the performance tradeoff.

**Comment by dneto0** [comment]
One of the sharp edges is that space available for push-constants can be very limited, e.g. 128bytes is the minimum required to be supported by a Vulkan implementation.

**Comment by Kangz** [comment]
Closing in favor of the ["push constant"](https://github.com/gpuweb/gpuweb/blob/b1882e76628ab6013fd8720b3f32391e913cb28f/proposals/push-constants.md) proposal already in the repo and being worked on.
